core/server/master/src/main/java/alluxio/master/throttle/ServerIndicator.java

```diff
@@ -0,0 +1,270 @@
+   * @param heapUsed the used heap
+   * @param cpuLoad the cpu load
+   * @param totalJVMPauseTimeMS the total JVM pause time since previous PIT
+   * @param pitTotalJVMPauseTimeMS the pit total JVM pause time
+   * @param rpcQueueSize the pit rpc queue size
+   * @param snapshotTimeMS the pit time
+   */
+  public ServerIndicator(long directMemUsed, long heapMax, long heapUsed, double cpuLoad,
+      long totalJVMPauseTimeMS, long pitTotalJVMPauseTimeMS, long rpcQueueSize,
+      long snapshotTimeMS) {
+    mDirectMemUsed = directMemUsed;
+    mHeapMax = heapMax;
+    mHeapUsed = heapUsed;
+    mCpuLoad = cpuLoad;
+    mTotalJVMPauseTimeMS = totalJVMPauseTimeMS;
+    mPITTotalJVMPauseTimeMS = pitTotalJVMPauseTimeMS;
+    mRpcQueueSize = rpcQueueSize;
+    mPitTimeMS = snapshotTimeMS;
+  }
+
+  /**
+   * @param serverIndicator the server indicator
+   */
+  public ServerIndicator(ServerIndicator serverIndicator) {
+    this(serverIndicator, 1);
+  }
+
+  /**
+   * The scaled server indicator according to the multiple.
+   *
+   * @param serverIndicator the base server indicator
+   * @param multiple the times
+   */
+  public ServerIndicator(ServerIndicator serverIndicator, int multiple) {
+    Preconditions.checkNotNull(serverIndicator, "serverIndicator");
+    mDirectMemUsed = serverIndicator.mDirectMemUsed * multiple;
+    mHeapMax = serverIndicator.mHeapMax;
+    mHeapUsed = serverIndicator.mHeapUsed * multiple;
+    mCpuLoad = serverIndicator.mCpuLoad * multiple;
+    mTotalJVMPauseTimeMS = serverIndicator.mTotalJVMPauseTimeMS * multiple;
+    mPITTotalJVMPauseTimeMS = serverIndicator.mPITTotalJVMPauseTimeMS;
+    mRpcQueueSize = serverIndicator.mRpcQueueSize * multiple;
+    mPitTimeMS = serverIndicator.mPitTimeMS;
+  }
+
+  /**
+   * @return current total JVM pause time
+   */
+  public static long getSystemTotalJVMPauseTime() {
+    if (Configuration.getBoolean(PropertyKey.MASTER_JVM_MONITOR_ENABLED)) {
+      return (long) (MetricsSystem.METRIC_REGISTRY
+          .gauge(MetricsMonitorUtils.ServerGaugeName.TOTAL_EXTRA_TIME, null).getValue());
+    }
+    return 0;
+  }
+
+  /**
+   * creates server indicator based on current status.
+   *
+   * @param prevTotalJVMPauseTime the previous total JVM pause time
+   * @return the server indicator
+   */
+  public static ServerIndicator createFromMetrics(long prevTotalJVMPauseTime) {
+    long pitTotalJVMPauseTime = 0;
+    long totalJVMPauseTime = 0;
+    long rpcQueueSize = 0;
+
+    if (Configuration.getBoolean(PropertyKey.MASTER_JVM_MONITOR_ENABLED)) {
+      pitTotalJVMPauseTime = (long) (MetricsSystem.METRIC_REGISTRY
+          .gauge(MetricsMonitorUtils.ServerGaugeName.TOTAL_EXTRA_TIME, null).getValue());
+      totalJVMPauseTime = pitTotalJVMPauseTime - prevTotalJVMPauseTime;
+    }
+    try {
+      rpcQueueSize = (long) (MetricsSystem.METRIC_REGISTRY
+          .gauge(MetricsMonitorUtils.ServerGaugeName.RPC_QUEUE_LENGTH, null).getValue());
+    } catch (NullPointerException e) {
```

Comment:
maybe use Optional.ofNullalbe and then map to avoid creating a null pointer exception.

Reply:
No, just a long, not Long

Reply:
I was thinking something like
`long rpcQueueSize = Optional.ofNullable((Long) MetricsSystem.METRIC_REGISTRY.gauge(MetricsMonitorUtils.ServerGaugeName.RPC_QUEUE_LENGTH, null).getValue()).orElse(0L);`

Reply:
It will throw an exception if not registered

Reply:
will this capture the exception?

Reply:
Oh ok I see, it is because it tries to register a new metric, but a null function is provided.

Now I am actually wondering if there is a bug here:
https://github.com/Alluxio/alluxio/blob/master/core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java#L522
It looks like after the server loses primary, then regains it, it will not register the new executor service for the metric, so it will not get the new value. What do you think @jenoudet ?